Change copy, enable sync all behind a user tag

artemruts opened the pull request:
### Description

These changes slightly change the copy and  enables `Sync all repositories` radio button when the user has the `AllowUserExternalServicePrivate` tag. Radio button are behind the shimmer effect now.

### Details

1. `AllowUserExternalServicePrivate` tag controls minor copy changes and enables the `Sync all repositories` radio button

2.  `Sync all repositories` radio button

Repo selection radio buttons had to be placed behind a shimmer.  
When the private code is enabled we need to determine if the user selected all of the affiliate repos, and if we should set `Sync all repositories` radio button as active. To do that we need to wait for two queries to complete:
    
  * `ExternalServices` - to check if code hosts are configure and to build a list of all synced (already selected) affiliated repos;
  * `AffiliatedRepositories` - to determine how many affiliated repos user has.
  
Queries need to happen serially because `AffiliatedRepositories` endpoint can only be queried when the code host(s) are configured. Executing queries serially gives us control wether we should send the `AffiliatedRepositories` request. Serial requests create latency and that's why the shimmer was modified and moved.


### How to test

Using `psql` connect to the `sourcegraph` (most likely the default) db and execute the following query:

1. Set user tag
a. adding user `AllowUserExternalServicePrivate` tag:
    
    ```sql
    update users set tags = array_append(tags, 'AllowUserExternalServicePrivate') where username = '<your username>';
    ```
    
    https://user-images.githubusercontent.com/1319181/116275902-05b77c80-a752-11eb-96db-01d3bf9c6e6c.mov
    
    
    b. removing user tag:
    
    ```sql
    update users set tags = array_remove(tags, 'AllowUserExternalServicePrivate') where username = '<your username>';
    ```
2.  Use personal access token to add new code host connection
3. Navigate to `Settings` → `Repositories` page

### Demos

<details>
  <summary>📹 &nbsp;Detailed demos. Same functionality, outdated shimmer</summary>

<br/>

1. Private code _enabled_, _all_ repos we synced

https://user-images.githubusercontent.com/1319181/115652302-050a8a80-a2fb-11eb-852d-4772e1ac4460.mov

2. Private code _enabled_, _some_ repos we synced

https://user-images.githubusercontent.com/1319181/115652499-62064080-a2fb-11eb-9981-676f98cacd7c.mov

3. Private code _enabled_, _none_ we synced

https://user-images.githubusercontent.com/1319181/115653158-a5ad7a00-a2fc-11eb-82d8-55eb17cb8741.mov

3. Private code _disabled_, _all public_ repos we synced

https://user-images.githubusercontent.com/1319181/115652795-e789f080-a2fb-11eb-90d6-be589ad38817.mov

4. Private code _disabled_, _some public_ repos we synced

https://user-images.githubusercontent.com/1319181/115652862-0ee0bd80-a2fc-11eb-9501-1a1348c22b7e.mov

5. Private code _disabled_, _none public_ we synced

https://user-images.githubusercontent.com/1319181/115653265-d9889f80-a2fc-11eb-982b-f9b218629c1c.mov

</details>

### Questions

- [ ] is it enough to put private repos behind the `AllowUserExternalServicePrivate` tag?

Page is accessible only by signed-in users with `AllowUserExternalServicePublic` tag. Do we need anything special for Admins?

artemruts commented:
@quinnkeast do we need to adjust copy in other way?

ryanslade commented:
I think we should make it clearer that when someone selects "Sync all repositories" it will sync repos from organisations they're a member of too. I think a lot of people who's personal GitHub account is also used for work might not expect that they'll be sending all their private work repos to SG too.

@quinnkeast WDYT?

quinnkeast commented:
> I think we should make it clearer that when someone selects "Sync all repositories" it will sync repos from organisations they're a member of too. I think a lot of people who's personal GitHub account is also used for work might not expect that they'll be sending all their private work repos to SG too.
> 
> @quinnkeast WDYT?

I think this will depend on the outcome of [this Slack thread](https://sourcegraph.slack.com/archives/CHPC7UX16/p1619080469216800?thread_ts=1619022613.197900&cid=CHPC7UX16). If we only allow OAuth, then there's a higher degree of inherent transparency around what can be synced, including the organizations.

quinnkeast commented:
@artemruts thanks for sharing the recordings! Couple of thoughts/questions:

- Shimmer for loading is a good solution. Thanks for adding this! Although it will cause a bit of "jump," I think we should remove the shimmer for the alert itself if possible, because that's a "temporary" rather than permanent element in the UI, so the skeleton doesn't make sense there.
- Based on current info and planning, we should remove/hide the option to "Sync all from selected organizations or users" and only show "Sync all" and "Sync selected." While I think this will come eventually, I don't think we're going to include this in the scope right now.

artemruts commented:
> @artemruts thanks for sharing the recordings! Couple of thoughts/questions:
> 
> * Shimmer for loading is a good solution. Thanks for adding this! Although it will cause a bit of "jump," I think we should remove the shimmer for the alert itself if possible, because that's a "temporary" rather than permanent element in the UI, so the skeleton doesn't make sense there.
> * Based on current info and planning, we should remove/hide the option to "Sync all from selected organizations or users" and only show "Sync all" and "Sync selected." While I think this will come eventually, I don't think we're going to include this in the scope right now.

Addressed everything, changed shimmer a bit too:


https://user-images.githubusercontent.com/1319181/115790867-6f700900-a395-11eb-9629-1351e224e237.mov

ryanslade commented:
> p.s. In private code mode all of the radio buttons are enabled and there's no "coming soon"

I think we decided that we don't want to enable the "sync all" quite yet. We're worried that if we enable that with private code enabled we risk users leaking private org code they have access to by not paying close attention (or even being aware of) all the code their token has access to.

cc @christinaforney

ryanslade commented:
Apart from that, it all looks great!

quinnkeast commented:
>I think we decided that we don't want to enable the "sync all" quite yet. We're worried that if we enable that with private code enabled we risk users leaking private org code they have access to by not paying close attention (or even being aware of) all the code their token has access to.

@ryanslade to be sure: this would only be accessible to Sourcegraph team members until private code launches in June. Do we still have this concern if this is the plan?

artemruts commented on client/web/src/user/settings/repositories/UserSettingsManageRepositoriesPage.tsx:
```diff
@@ -141,6 +146,8 @@ export const UserSettingsManageRepositoriesPage: React.FunctionComponent<Props>
         telemetryService.logViewEvent('UserSettingsRepositories')
     }, [telemetryService])
 
+    const ALLOW_PRIVATE_CODE = user.tags.includes('AllowUserExternalServicePrivate')
```
@asdine Do you think this check will be enough?

artemruts commented:
@asdine I was thinking about `private code` vs `sync all` functionality and I want to split them into two separate feature toggles.

I think those are two unconnected things:
* `private code` - is for code hosts to check if they should also include private repos
* `sync all` - means sync all `public` or sync all `public + private` repos; just a "command" for the code host to sync "everything they have".

If we have both `AllowUserExternalServicePrivate` and `AllowUserExternalServiceSyncAll` user tags this will allow for more gradual rollout:

1. `{..., AllowUserExternalServicePrivate}` - users can only individually select public and private repos
2. `{..., AllowUserExternalServicePrivate, AllowUserExternalServiceSyncAll}` - users can select individually as well as all of the private and public repos

___

I think this will require adding a `AllowUserExternalServiceSyncAll` user tag in [set_external_service_repos.go](https://github.com/sourcegraph/sourcegraph/blob/main/cmd/frontend/graphqlbackend/set_external_service_repos.go#L58).

I already added check for `AllowUserExternalServiceSyncAll` in this PR.

WDYT?

artemruts commented:
> @asdine I was thinking about `private code` vs `sync all` functionality and I want to split them into two separate feature toggles.
> 
> I think those are two unconnected things:
> 
> * `private code` - is for code hosts to check if they should also include private repos
> * `sync all` - means sync all `public` or sync all `public + private` repos; just a "command" for the code host to sync "everything they have".
> 
> If we have both `AllowUserExternalServicePrivate` and `AllowUserExternalServiceSyncAll` user tags this will allow for more gradual rollout:
> 
> 1. `{..., AllowUserExternalServicePrivate}` - users can only individually select public and private repos
> 2. `{..., AllowUserExternalServicePrivate, AllowUserExternalServiceSyncAll}` - users can select individually as well as all of the private and public repos
> 
> I think this will require adding a `AllowUserExternalServiceSyncAll` user tag in [set_external_service_repos.go](https://github.com/sourcegraph/sourcegraph/blob/main/cmd/frontend/graphqlbackend/set_external_service_repos.go#L58).
> 
> I already added check for `AllowUserExternalServiceSyncAll` in this PR.
> 
> WDYT?

Was discussed offline.

unknwon reviewed (approved):
Given this is behind a feature flag, the impact is minimized and I do not see a reason to block from merging and test live.
